File: huffman.c
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

// Workarounds for webAssembly
#ifdef __EMSCRIPTEN__
#include <emscripten.h>

EMSCRIPTEN_KEEPALIVE
void *my_malloc(uint32_t size)
{
    return malloc(size);
};
EMSCRIPTEN_KEEPALIVE
void my_free(void *cur)
{
    return free(cur);
};
#else
#include <stdio.h>
#include <stdarg.h>
#define EMSCRIPTEN_KEEPALIVE
#endif

void debug(char *fmt, ...)
{
#ifndef __EMSCRIPTEN__
    va_list args;
    va_start(args, fmt);
    // vprintf(fmt,args);
    va_end(args);
#endif
}

/** left&right or byte */
struct HuffmanNode
{
    short int byte;
    uint32_t count;
    short int nextIdx;
    short int leftIdx;
    short int rightIdx;
};

// Just random values for header
char magic[] = {85, 92, 110, 65};

struct StreamNode
{
    unsigned char bitLength; // Ranges are 1..255
    unsigned char bits[32];
};
void writeBitToStreamNode(struct StreamNode *node, char bitIdx, char bit)
{

    unsigned char currentBitStreamNode = bitIdx & 7;
    unsigned char currentByteStreamNode = bitIdx >> 3;
    debug("Writing bit bitIdx=%i, bit=%i byte=%i \n", bitIdx, currentBitStreamNode, currentByteStreamNode);
    if (bit > 0)
    {
        char targetMask = 1 << (7 - currentBitStreamNode);
        node->bits[currentByteStreamNode] = node->bits[currentByteStreamNode] | targetMask;
    }
    else
    {
        char targetMask = 255 - (1 << (7 - currentBitStreamNode));
        node->bits[currentByteStreamNode] = node->bits[currentByteStreamNode] & targetMask;
    };
};

void goTreeForStream(struct HuffmanNode *list, struct StreamNode *bytes, int idx, struct StreamNode *path)
{
    debug("In node %i. path len = %i \n", idx, path->bitLength);
    if (list[idx].leftIdx != -1 && list[idx].rightIdx != -1)
    {
        debug("In internal node %i. left=%i right=%i \n", idx, list[idx].leftIdx, list[idx].rightIdx);
        struct StreamNode *left = path;
        struct StreamNode *right = (struct StreamNode *)(malloc(sizeof(struct StreamNode)));
        memcpy(right, left, sizeof(struct StreamNode));
        writeBitToStreamNode(left, left->bitLength, 0);
        writeBitToStreamNode(right, right->bitLength, 1);
        debug("Bits left=%i right=%i\n", left->bitLength, right->bitLength);
        left->bitLength++;
        right->bitLength++;
        debug("After ++ left=%i right=%i\n", left->bitLength, right->bitLength);
        goTreeForStream(list, bytes, list[idx].leftIdx, left);
        goTreeForStream(list, bytes, list[idx].rightIdx, right);
    }
    else
    {
        debug("In leaf node %i. left=%i right=%i \n", idx, list[idx].leftIdx, list[idx].rightIdx);
        if (idx > 256)
        {
            debug("Internal error, wrong idx for leaf node\n");
        }
        memcpy(&bytes[idx], path, sizeof(struct StreamNode));
        free(path);
    };
};

void writeBit(unsigned char *out, uint32_t *outlen, unsigned char *currentBit, char bit)
{
    if (bit > 0)
    {
        char targetMask = 1 << (7 - *currentBit);
        // debug("mask=%x\n", targetMask);
        // debug("out before=%x\n", out[*outlen]);
        out[*outlen] = out[*outlen] | targetMask;
        // debug("out after=%x\n", out[*outlen]);
    }
    else
    {
        char targetMask = 255 - (1 << (7 - *currentBit));
        out[*outlen] = out[*outlen] & targetMask;
    }
    *currentBit = *currentBit + 1;
    if (*currentBit >= 8)
    {
        *currentBit = 0;
        *outlen += 1;
    }
};

void writeHeaderNode(struct HuffmanNode *list, unsigned char *out, uint32_t *outlen, unsigned char *currentBit, int idx)
{
    debug("Writing header node %i. left=%i right=%i \n", idx, list[idx].leftIdx, list[idx].rightIdx);
    if (list[idx].leftIdx != -1 && list[idx].rightIdx != -1)
    {
        debug("It is Internal\n");
        writeBit(out, outlen, currentBit, 0);
        writeHeaderNode(list, out, outlen, currentBit, list[idx].leftIdx);
        writeHeaderNode(list, out, outlen, currentBit, list[idx].rightIdx);
    }
    else
    {
        debug("It is Leaf byte=%u\n", list[idx].byte);
        writeBit(out, outlen, currentBit, 1);
        for (int i = 0; i < 8; i++)
        {
            char bit = list[idx].byte >> (7 - i) & 1;
            writeBit(out, outlen, currentBit, bit);
        }
    }
};

EMSCRIPTEN_KEEPALIVE
unsigned char *huffman_encode(unsigned char *in, uint32_t len, uint32_t *outlen)
{
    debug("Creating buffer for counts and counting\n");
    uint32_t counts[256] = {0};
    for (uint32_t i = 0; i < len; i++)
    {
        counts[in[i]]++;
    };

    struct HuffmanNode *list = (struct HuffmanNode *)malloc(sizeof(struct HuffmanNode) * 511);
    // debug("Size of node is %i\n", sizeof(struct HuffmanNode)); // Lol, it is aligned

    debug("Building initial list\n");
    for (int i = 0; i < 256; i++)
    {

        list[i].byte = (unsigned char)i;
        list[i].count = counts[i];
        list[i].leftIdx = -1;
        list[i].rightIdx = -1;
        list[i].nextIdx = i < 255 ? i + 1 : -1;
    }

    int listIdx = 0;
    int freeIdx = 256;
    debug("Transforming list into tree\n");
    while (1)
    {
        int min1idx = -1;
        int min2idx = -1;
        int currentIdx = listIdx;
        if (list[listIdx].nextIdx == -1)
        {
            // This should never happen because we check and break in the end of this while loop
            break;
        }
        // Two passes to find two minimums. Not optimal
        while (currentIdx != -1)
        {
            if (min1idx == -1 || list[currentIdx].count < list[min1idx].count)
            {
                min1idx = currentIdx;
            }
            currentIdx = list[currentIdx].nextIdx;
        }
        currentIdx = listIdx;
        while (currentIdx != -1)
        {
            if (min2idx == -1 || list[currentIdx].count < list[min2idx].count)
            {
                if (min1idx != currentIdx)
                {
                    min2idx = currentIdx;
                }
            }
            currentIdx = list[currentIdx].nextIdx;
        }
        if (list[min1idx].leftIdx != -1)
        {
            debug("MIN1 leaf left=%i right=%i min1idx=%i\n", list[min1idx].leftIdx, list[min1idx].rightIdx, min1idx);
        }
        else
        {
            debug("MIN1 byte=%i min1idx=%i\n", list[min1idx].byte, min1idx);
        };
        if (list[min2idx].leftIdx != -1)
        {
            debug("MIN2 leaf left=%i right=%i min1idx=%i\n", list[min2idx].leftIdx, list[min2idx].rightIdx, min2idx);
        }
        else
        {
            debug("MIN2 byte=%i min2idx=%i\n", list[min2idx].byte, min2idx);
        };

        list[freeIdx].count = list[min1idx].count + list[min2idx].count;
        list[freeIdx].leftIdx = min1idx;
        list[freeIdx].rightIdx = min2idx;
        list[freeIdx].nextIdx = -1;

        // Removing items from list
        debug("Removing min1 from list\n");
        int previousIdx = -1;
        currentIdx = listIdx;
        while (currentIdx != -1)
        {
            if (currentIdx == min1idx)
            {
                if (previousIdx == -1)
                {
                    listIdx = list[currentIdx].nextIdx;
                }
                else
                {
                    list[previousIdx].nextIdx = list[currentIdx].nextIdx;
                }
                break;
            }
            previousIdx = currentIdx;
            currentIdx = list[currentIdx].nextIdx;
        }
        debug("Removing min2 from list\n");
        previousIdx = -1;
        currentIdx = listIdx;
        while (currentIdx != -1)
        {
            if (currentIdx == min2idx)
            {
                if (previousIdx == -1)
                {
                    listIdx = list[currentIdx].nextIdx;
                }
                else
                {
                    list[previousIdx].nextIdx = list[currentIdx].nextIdx;
                }
                break;
            }
            previousIdx = currentIdx;
            currentIdx = list[currentIdx].nextIdx;
        }

        debug("Add new item to list\n");
        if (listIdx == -1)
        {
            debug("List is null, tree is built\n");
            break;
        }
        else
        {
            currentIdx = listIdx;
            while (list[currentIdx].nextIdx != -1)
            {

                currentIdx = list[currentIdx].nextIdx;
            }
            list[currentIdx].nextIdx = freeIdx;
        }

        int i = 0;
        currentIdx = listIdx;
        while (currentIdx != -1)
        {
            i++;
            currentIdx = list[currentIdx].nextIdx;
        }
        debug("List have %i items\n", i);
        freeIdx++;
    };

    // We did 255 steps in the previous loop, each step was removing
    //  two items from list and adding one. So, it always will be 511 nodes
    /*
    int nodes_count = 0;
    void count_nodes(struct HuffmanNode * pos)
    {
        nodes_count++;
        if (pos->left != NULL)
        {
            count_nodes(pos->left);
        };
        if (pos->right != NULL)
        {
            count_nodes(pos->right);
        };
    };
    count_nodes(list);
    debug("Tree have %i nodes\n", nodes_count);
    */
    //

    debug("Creating bytes structure\n");
    struct StreamNode *bytes = (struct StreamNode *)(malloc(sizeof(struct StreamNode) * 256));
    struct StreamNode *initial = (struct StreamNode *)(malloc(sizeof(struct StreamNode)));
    initial->bitLength = 0;
    for (unsigned char i = 0; i < 32; i++)
    {
        initial->bits[i] = 0;
    };

    goTreeForStream(list, bytes, 510, initial);

    // TODO: it will overflow on big files
    //   need to implelement overflow safe-multiplication
    uint32_t total_stream_size_bits = 0;
    for (int i = 0; i < 256; i++)
    {
        total_stream_size_bits += counts[i] * bytes[i].bitLength;
    };
    uint32_t total_stream_bytes = (total_stream_size_bits >> 3) +
                                  (((total_stream_size_bits & 7) != 0) ? 1 : 0);
    unsigned char *out = malloc(total_stream_bytes + 328);

    debug("Initializing header\n");
    *outlen = 0;
    out[0] = magic[0];
    out[1] = magic[1];
    out[2] = magic[2];
    out[3] = magic[3];
    *outlen += 4;

    *(uint32_t *)&out[4] = len;
    *outlen += sizeof(uint32_t);

    unsigned char currentBit = 0;

    debug("Writing header\n");
    writeHeaderNode(list, out, outlen, &currentBit, 510);
    debug("Header is written,outlen=%lu bit=%i\n", *outlen, currentBit);
    if (*outlen != 327 || currentBit != 7)
    {
        debug("Internal error with outlen and bits: %i %i\n", *outlen, currentBit);
        return NULL;
    }
    writeBit(out, outlen, &currentBit, 0);
    free(list);
    debug("Built a list of bytes, start to write byte stream\n");

    for (uint32_t i = 0; i < len; i++)
    {
        struct StreamNode *node = &bytes[in[i]];
        //debug("Byte idx = %lu, val = %x\n", i, in[i]);
        //debug("Node bitLen=%i\n", node->bitLength);
        //debug("Node bytes %x %x %x %x", node->bits[0], node->bits[1], node->bits[2], node->bits[3]);
        //return NULL;
        for (unsigned char bitPos = 0; bitPos < node->bitLength; bitPos++)
        {

            unsigned char bitBitPos = bitPos & 7;
            unsigned char bitBytePos = bitPos >> 3;
            unsigned char bitValue = node->bits[bitBytePos] >> (7 - bitBitPos) & 1;
            writeBit(out, outlen, &currentBit, bitValue);
        }
    };
    debug("Input written\n");
    free(bytes);
    while (currentBit != 0)
    {
        writeBit(out, outlen, &currentBit, 0);
    };

    debug("Done\n");
    return out;
};

EMSCRIPTEN_KEEPALIVE
unsigned char *huffman_decode(unsigned char *in, uint32_t *outlen)
{
    debug("Checking header magic\n");
    if (in[0] != magic[0] ||
        in[1] != magic[1] || in[2] != magic[2] || in[3] != magic[3])
    {
        debug("Header is not correct");
        return NULL;
    }

    *outlen = *(uint32_t *)&in[4];

    //

    //
    debug("Creating buffer for counts and counting\n");
}

File: huffman.ts
function debug(str: string) {
    // console.info(str);
}
export function huffmanEncode(input: Uint8Array) {
    debug("Calculating bytes frequency");
    const counts = new Array(256).fill(0);
    for (const byte of input) {
        counts[byte]++;
    }
    interface TreeAndList {
        byte: number;
        next?: TreeAndList;
        left?: TreeAndList;
        right?: TreeAndList;
        idx?: number; // Index for internal node to pack into header
        count: number;
    }
    debug("Building initial list");
    // This is not the most efficient way
    //  We know what each Huffman tree will hold 511 nodes,
    //  so we can initiate an array with 511 items and use
    //  array indexes instead of pointers
    let tree: TreeAndList | undefined = undefined;
    let last: TreeAndList | undefined = undefined;
    for (let i = 0; i < 256; i++) {
        const byte = i;
        const count = counts[i];
        const item: TreeAndList = {
            byte,
            count
        };
        if (!last) {
            tree = item;
            last = item;
        } else {
            last.next = item;
        }
        last = item;
    }
    if (!tree) {
        throw new Error("Internal error");
    }
    let lastInternalNodeIdx = 0;
    while (true) {
        if (!tree.next) {
            throw new Error("Internal error");
        }
        let current: TreeAndList | undefined = tree;
        let min1: TreeAndList | undefined = undefined;
        let min2: TreeAndList | undefined = undefined;

        // Find a first minimum in list
        while (current) {
            if (!min1 || current.count < min1.count) {
                min1 = current;
            }
            current = current.next;
        }

        // Find a second minimum in list
        current = tree;
        while (current) {
            if (!min2 || current.count < min2.count) {
                if (current !== min1) {
                    min2 = current;
                }
            }
            current = current.next;
        }
        if (!min1) {
            throw new Error("Internal error: min1 not found");
        }
        if (!min2) {
            throw new Error("Internal error: min2 not found");
        }
        // Create a new iternal node
        lastInternalNodeIdx++;
        const node: TreeAndList = {
            byte: 0,
            count: min1.count + min2.count,
            left: min1,
            right: min2,
            idx: lastInternalNodeIdx,
            next: undefined
        };

        // Remove min1 from list
        current = tree;
        last = undefined;
        while (current) {
            if (current === min1) {
                if (!last) {
                    tree = current.next;
                } else {
                    last.next = current.next;
                }
            }
            last = current;
            current = current.next;
        }

        // Remove min2 from list
        current = tree;
        last = undefined;
        while (current) {
            if (current === min2) {
                if (!last) {
                    tree = current.next;
                } else {
                    last.next = current.next;
                }
            }
            last = current;
            current = current.next;
        }

        // Add a new item into list
        if (!tree) {
            // That was last pair. Tree is new node
            tree = node;
            break;
        } else {
            current = tree;
            while (current) {
                if (!current.next) {
                    current.next = node;
                    break;
                } else {
                    current = current.next;
                }
            }
        }
    }
    debug("Tree created");

    debug(`Creating byte helper`);
    type HelperNode = number[];

    const helpers: HelperNode[] = [];
    function goTree(node: TreeAndList, path: HelperNode) {
        if (node.left && node.right) {
            const left = path.slice();
            const right = path.slice();
            left.push(0);
            right.push(1);
            goTree(node.left, left);
            goTree(node.right, right);
        } else {
            helpers[node.byte] = path;
        }
    }
    goTree(tree, []);

    let totalStreamBitLength = 0;
    for (let i = 0; i < 256; i++) {
        totalStreamBitLength += helpers[i].length * counts[i];
    }
    const totalStreamByteLength =
        (totalStreamBitLength >> 3) +
        ((totalStreamBitLength & 7) === 0 ? 0 : 1);
    const totalLength = 328 + totalStreamByteLength;
    const output = new Uint8Array(totalLength);
    output[0] = 85;
    output[1] = 92;
    output[2] = 110;
    output[3] = 65;
    output[4] = input.byteLength & 255;
    output[5] = (input.byteLength >> 8) & 255;
    output[6] = (input.byteLength >> 16) & 255;
    output[7] = (input.byteLength >> 24) & 255;

    // The tree have 256 leaf nodes and 255 internal nodes
    // Bit: 0 - internal, 1 - leaf
    //  if internal, then two childs go next
    //  if leaf, then byte go next

    let currentByte = 8;
    let currentBit = 0;
    function writeBit(bit: 0 | 1) {
        if (bit === 1) {
            const targetMask = 1 << (7 - currentBit);
            output[currentByte] = output[currentByte] | targetMask;
        } else {
            const targetMask = 255 - (1 << (7 - currentBit));
            output[currentByte] = output[currentByte] & targetMask;
        }
        currentBit++;
        if (currentBit >= 8) {
            currentBit = 0;
            currentByte++;
        }
    }

    function writeHeader(current: TreeAndList) {
        if (current.left && current.right) {
            // it is a internal node
            debug(
                `Writing internal node idx=${
                    current.idx !== undefined ? current.idx + 255 : "err"
                }`
            );
            if (current.idx === undefined) {
                throw new Error("Internal error");
            }
            writeBit(0);
            writeHeader(current.left);
            writeHeader(current.right);
        } else {
            debug(`Writing leaf node byte=${current.byte}`);
            writeBit(1);
            for (let i = 0; i < 8; i++) {
                const bit = (current.byte >> (7 - i)) & 1;
                writeBit(bit ? 1 : 0);
            }
        }
    }
    writeHeader(tree);
    debug(`Header is written, byte=${currentByte} bit=${currentBit}`);
    if (currentByte !== 327 || currentBit !== 7) {
        throw new Error(`Internal error: ${currentByte} ${currentBit}`);
    }
    writeBit(0);

    debug("Writing data");

    //debug(`currentbyte=${currentByte} currentBit=${currentBit}`);

    for (const byte of input) {
        const bits = helpers[byte];
        //debug(byte, bits, bits.length);
        //process.exit(0);
        if (!bits) {
            throw new Error("Internal error");
        }
        for (const bit of bits) {
            writeBit(bit ? 1 : 0);
        }
    }

    while (currentBit > 0) {
        writeBit(0);
    }
    if (currentByte !== totalLength) {
        throw new Error(`Internal error ${currentByte} ${totalLength}`);
    }
    return output;
    //debug(`Header bit size = ${bitPos}`);
}

export function huffmanDecode(input: Uint8Array) {
    if (
        input[0] !== 85 ||
        input[1] !== 92 ||
        input[2] !== 110 ||
        input[3] !== 65
    ) {
        throw new Error("Wrong header");
    }
    const size =
        input[4] + (input[5] << 8) + (input[6] << 16) + (input[7] << 24);
    debug(`Output size = ${size}`);
    interface HuffmanNode {
        left?: HuffmanNode;
        right?: HuffmanNode;
        byte?: number;
    }
    const tree: HuffmanNode = {};
    let bytePos = 8;
    let bitPos = 0;
    function readBit() {
        const bit = (input[bytePos] >> (7 - bitPos)) & 1;
        bitPos++;
        if (bitPos >= 8) {
            bytePos++;
            bitPos = 0;
        }
        return bit;
    }
    function readTree(node: HuffmanNode) {
        const type = readBit();
        if (type === 0) {
            debug("Got internal node");
            const left: HuffmanNode = {};
            const right: HuffmanNode = {};
            node.left = left;
            node.right = right;
            readTree(left);
            readTree(right);
        } else {
            let byte = 0;
            for (let i = 0; i < 8; i++) {
                const bit = readBit();
                byte += bit > 0 ? 2 ** (7 - i) : 0;
            }
            debug(`Got leaf node byte=${byte}`);
            node.byte = byte;
        }
    }
    readTree(tree);
    debug(`Tree read done`);
    if (bytePos !== 327 || bitPos !== 7) {
        throw new Error("Internal error");
    }
    const padder = readBit();
    if (padder !== 0) {
        throw new Error("Wrong stream");
    }
    const out = new Uint8Array(size);
    let currentByte = 0;
    while (currentByte < size) {
        let node: HuffmanNode = tree;
        while (node.byte === undefined) {
            const route = readBit();
            if (route === 0) {
                if (node.left === undefined) {
                    throw new Error("Internal error");
                }
                node = node.left;
            } else {
                if (node.right === undefined) {
                    throw new Error("Internal error");
                }
                node = node.right;
            }
        }
        out[currentByte] = node.byte;
        currentByte++;
    }
    debug("Written");
    return out;
}

/*

ts-node -T huffman.ts

*/
